### PR TITLE
Thread safety for LinkFileWriter.

### PR DESCRIPTION
**Pull request by vwxyzh**
@chenkennt @ansyral @hellosnow @vicancy @qinezh @superyyrrzz

### CODE REVIEW

**Comment by hellosnow** [comment]
:shipit:

**Comment by superyyrrzz** [comment]
:shipit:

File: src/Microsoft.DocAsCode.Common/FileAbstractLayer/LinkFileWriter.cs [hunk 1]
```diff
@@ -20,32 +20,44 @@ public LinkFileWriter(string outputFolder)
         public override void Copy(PathMapping sourceFilePath, RelativePath destFilePath)
         {
             var key = destFilePath.GetPathFromWorkingFolder();
-            _mapping[key] = new PathMapping(key, sourceFilePath.PhysicalPath)
+            var pm = new PathMapping(key, sourceFilePath.PhysicalPath)
             {
                 Properties = sourceFilePath.Properties,
             };
+            lock (_mapping)
```
Comment (ansyral) [inline src/Microsoft.DocAsCode.Common/FileAbstractLayer/LinkFileWriter.cs]: >lock (_mapping) [](start = 12, length = 15)

question: what's the advantage over directly using ConcurentDictionary?

Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Common/FileAbstractLayer/LinkFileWriter.cs]: https://msdn.microsoft.com/en-us/library/dd287131(v=vs.110).aspx#Anchor_1
> The enumerator returned from the dictionary is safe to use concurrently with reads and writes to the dictionary, however it does not represent a moment-in-time snapshot of the dictionary. The contents exposed through the enumerator may contain modifications made to the dictionary after GetEnumerator was called.

**Pull request merged**